cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java

```diff
@@ -0,0 +1,161 @@
+   */
+  @Nullable
+  public static MainClassLoader createFromContext() {
+    return createFromContext(new FilterClassLoader.Filter() {
+      @Override
+      public boolean acceptResource(String resource) {
+        return false;
+      }
+
+      @Override
+      public boolean acceptPackage(String packageName) {
+        return false;
+      }
+    });
+  }
+
+  /**
+   * @return a new instance from the current context classloader or the system classloader. The returned
+   * {@link MainClassLoader} will be the defining classloader for classes in the context classloader
+   * that the filter rejected. For classes that pass the filter, the defining classloader will be the original
+   * context classloader.
+   * It will return {@code null} if it is not able to create a new instance due to lack of classpath information.
+   */
+  @Nullable
+  public static MainClassLoader createFromContext(FilterClassLoader.Filter filter) {
+    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+    if (classLoader == null) {
+      classLoader = ClassLoader.getSystemClassLoader();
+    }
+
+    URL[] classpath;
+
+    if (classLoader instanceof URLClassLoader) {
+      classpath = ((URLClassLoader) classLoader).getURLs();
+    } else if (classLoader == ClassLoader.getSystemClassLoader()) {
+      classpath = getClassPath();
+    } else {
+      // No able to create a new MainClassLoader
+      return null;
+    }
+
+    ClassLoader filtered = new FilterClassLoader(classLoader, filter);
+    ClassLoader parent = new CombineClassLoader(classLoader.getParent(), Collections.singleton(filtered));
+    return new MainClassLoader(classpath, parent);
+  }
+
+  /**
+   * Creates a new instance for the following set of {@link URL}.
+   *
+   * @param urls the URLs from which to load classes and resources
+   * @param parent the parent classloader for delegation
+   */
+  public MainClassLoader(URL[] urls, ClassLoader parent) {
+    super(urls, parent);
+    this.datasetRewriter = new DatasetClassRewriter();
+    this.resourceLookup = ClassLoaders.createClassResourceLookup(this);
+    this.cache = new HashMap<>();
+  }
+
+  @Override
+  protected boolean needIntercept(String className) {
+    try {
+      return Classes.isSubTypeOf(className, DATASET_CLASS_NAME, resourceLookup, cache);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    return datasetRewriter.rewriteClass(className, input);
+  }
+
+  /**
+   * Returns the list of {@link URL} for the system ClassLoader.
+   */
+  private static URL[] getClassPath() {
+    List<URL> urls = new ArrayList<>();
+    ClassLoader sysClassLoader = ClassLoader.getSystemClassLoader();
+
+    if (sysClassLoader instanceof URLClassLoader) {
+      urls.addAll(Arrays.asList(((URLClassLoader) sysClassLoader).getURLs()));
```

Comment:
Actually this `if` is unnecessary since we already check in the `createFromContext` method that if the classloader is an instance of `URLClassLoader`, hence this condition will never happen. Removed the if.